### PR TITLE
Add Quint simple expression conversion and operdefs

### LINKED ISSUE #2351: Convert Quint expressions and operator defs to ApalacheIR

### PR DESCRIPTION
**Pull request by shonfeder**
Closes #2351

Support for basic expressions and operator defs are added here, and operator definitions, along with some supporting utilities.

### CODE REVIEW

File: tla-io/src/main/scala/at/forsyte/apalache/io/quint/Quint.scala [hunk 1]
```diff
@@ -1,18 +1,133 @@
...
+            case lam: QuintLambda => lambdaBodyAndParams(lam)
+            // Otherwise it's an operator with no params
+            case other => (expConverter(other), List())
+          }
+          exp.decl(name, body, typedParams: _*)
+        // TODO
+        case QuintConst(_, _, _)   => null
+        case QuintVar(_, _, _)     => null
+        case QuintAssume(_, _, _)  => null
+        case QuintTypeDef(_, _, _) => null
```
Comment (shonfeder) [inline tla-io/src/main/scala/at/forsyte/apalache/io/quint/Quint.scala:98]: I'll be picking these up in a followup.

**Review by bugarela — COMMENTED** [review]
I am unsure whether this is ready for review, as it is still a draft but you requested a review. I missed some tests to get a better sense of what is going on, but it's fine by me if you want to merge this first and work on tests later.

File: tla-io/src/main/scala/at/forsyte/apalache/io/quint/Quint.scala [hunk 2]
```diff
@@ -1,18 +1,136 @@
...
+    import QuintEx._
+    import QuintType._
+
+    // Construct Apalache IR expressions
+    val exp = new ScopedBuilder()
+
+    // Derive a OperParam from a paramter name and it's type.
+    //
+    // OperParams are required by the ScopedBuilder for building
+    // operators and consist of a the params name and its arity,
```
[the remark this answers is not in the document]
Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/quint/Quint.scala:59]:
```suggestion
    // operators and consist of the param's name and its arity,
```

[the remark this answers is not in the document]
Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/quint/Quint.scala:59]: an extra `a`! thanks.

**Comment by shonfeder** [comment]
I had meant to mark it as ready for review.

> I missed some tests to get a better sense of what is going on

I have been thinking about testing, and haven't yet hit on anything sensible other than integration tests run against json IR generated by quint: hardcoding ir fragments will quickly become impossible to read for insight and a maintenance burden when the quint ir changes.

I'm thinking the best might be to rig up integration tests that can generate quint IR from string inputs?

**Comment by bugarela** [comment]
> I'm thinking the best might be to rig up integration tests that can generate quint IR from string inputs?

Do you mean by using Quint's CLI to parse the string and obtain the JSON? That seems like a good idea for integration tests, although the Apalache IR side of the comparison wouldn't be that readable anyway

I think it would make sense to have unit tests for each translation case, but not worrying too much about the recursion/combination of elements. i.e. This minimal lambda QuintEx results in this AplacheIR.

**Comment by shonfeder** [comment]
i'll try that scope and see how it goes. thanks for the suggestion

**Comment by konnov** [comment]
> Do you mean by using Quint's CLI to parse the string and obtain the JSON? That seems like a good idea for integration tests, although the Apalache IR side of the comparison wouldn't be that readable anyway

We could produce JSON with Quint, feed it to the transpiler to obtain Apalache JSON, which we would feed to `apalache parse` to print a TLA+ file. So that test would get `.qnt` at its input and compare the produced `.tla` with the expected one. This kind of tests would be the easiest to debug.

**Review by konnov — APPROVED** [review]
I had a quick look. This is a great start! It's cool that you are using the typed builder! 👯‍♂️

**Comment by shonfeder** [comment]
> We could produce JSON with Quint, feed it to the transpiler to obtain Apalache JSON, which we would feed to apalache parse to print a TLA+ file. So that test would get .qnt at its input and compare the produced .tla with the expected one. This kind of tests would be the easiest to debug.

Oh yeah, that's the ideal integration testing setup.

**Comment by shonfeder** [comment]
thanks for the reviews!  i'll add minimal unit tests, as suggested by @bugarela, in a followup, so that it can get a proper fresh review, and i'll keep thinking on cost/benefits of various approaches to integration tests

**Pull request merged**